TestSparseHistogram integer divide by zero on ppc64le

Whilst updating the [golang-github-prometheus-client-golang](https://tracker.debian.org/pkg/golang-github-prometheus-client-golang) package in Debian, autopkgtest revealed that the `TestSparseHistogram` test panics with an integer divide by zero on ppc64el (ppc64le as GOARCH).
```
=== RUN   TestSparseHistogram
=== RUN   TestSparseHistogram/no_sparse_buckets
=== RUN   TestSparseHistogram/factor_1.1_results_in_schema_3
=== RUN   TestSparseHistogram/factor_1.2_results_in_schema_2
=== RUN   TestSparseHistogram/factor_4_results_in_schema_-1
--- FAIL: TestSparseHistogram (0.00s)
    --- PASS: TestSparseHistogram/no_sparse_buckets (0.00s)
    --- PASS: TestSparseHistogram/factor_1.1_results_in_schema_3 (0.00s)
    --- PASS: TestSparseHistogram/factor_1.2_results_in_schema_2 (0.00s)
    --- FAIL: TestSparseHistogram/factor_4_results_in_schema_-1 (0.00s)
panic: runtime error: integer divide by zero [recovered]
	panic: runtime error: integer divide by zero

goroutine 3103 [running]:
testing.tRunner.func1.2({0x434f00, 0x8244b0})
	/usr/lib/go-1.19/src/testing/testing.go:1396 +0x1fc
testing.tRunner.func1()
	/usr/lib/go-1.19/src/testing/testing.go:1399 +0x364
panic({0x434f00, 0x8244b0})
	/usr/lib/go-1.19/src/runtime/panic.go:884 +0x240
github.com/prometheus/client_golang/prometheus.(*histogramCounts).observe(0xc0001b4750, 0xc000061d18?, 0xb0710?, 0x0?)
	/tmp/autopkgtest-lxc.4sj9alys/downtmp/autopkgtest_tmp/_build/src/github.com/prometheus/client_golang/prometheus/histogram.go:636 +0x324
github.com/prometheus/client_golang/prometheus.(*histogram).observe(0xc0003d8000, 0x3fe0000000000000, 0x10?)
	/tmp/autopkgtest-lxc.4sj9alys/downtmp/autopkgtest_tmp/_build/src/github.com/prometheus/client_golang/prometheus/histogram.go:807 +0xa4
github.com/prometheus/client_golang/prometheus.(*histogram).Observe(0xc0003d8000, 0xc95125dbd?)
	/tmp/autopkgtest-lxc.4sj9alys/downtmp/autopkgtest_tmp/_build/src/github.com/prometheus/client_golang/prometheus/histogram.go:707 +0x5c
github.com/prometheus/client_golang/prometheus.TestSparseHistogram.func1(0xc0003ae820)
	/tmp/autopkgtest-lxc.4sj9alys/downtmp/autopkgtest_tmp/_build/src/github.com/prometheus/client_golang/prometheus/histogram_test.go:675 +0x35c
testing.tRunner(0xc0003ae820, 0xc0003aa470)
	/usr/lib/go-1.19/src/testing/testing.go:1446 +0x11c
created by testing.(*T).Run
	/usr/lib/go-1.19/src/testing/testing.go:1493 +0x36c
FAIL	github.com/prometheus/client_golang/prometheus	24.050s
```
All other tested platforms (amd64, arm64, arm, i386, s390x) pass the test.

Since this may in fact be a Go compiler bug, I will disable that test on ppc64le, but thought it would be prudent to at least raise the issue here, in case there is an easy fix.

cc: @beorn7

The following minimal snippet is enough to provoke the bug:
```go
package main

import (
        "fmt"

        "github.com/prometheus/client_golang/prometheus"
)

func main() {
        his := prometheus.NewHistogram(prometheus.HistogramOpts{
                NativeHistogramBucketFactor: 4,
        })

        his.Observe(42)

        fmt.Println("Ok")
}
```
Execution:
```
$ /usr/lib/go-1.19/bin/go run main.go
panic: runtime error: integer divide by zero

goroutine 1 [running]:
github.com/prometheus/client_golang/prometheus.(*histogramCounts).observe(0xc00014a090, 0xc00011fe28?, 0xb4530?, 0x0?)
        /home/dswarbrick/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/histogram.go:636 +0x324
github.com/prometheus/client_golang/prometheus.(*histogram).observe(0xc000089c70, 0x4045000000000000, 0x0?)
        /home/dswarbrick/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/histogram.go:807 +0xa4
github.com/prometheus/client_golang/prometheus.(*histogram).Observe(0xc000089c70, 0x0?)
        /home/dswarbrick/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/histogram.go:707 +0x5c
main.main()
        /home/dswarbrick/foo/main.go:14 +0x80
exit status 2
```
It doesn't seem to matter what number we `Observe()`, and the panic occurs if `NativeHistogramBucketFactor` >= 4.

@beorn7 The bitshift on line 635 in prometheus/histogram.go is indeed where it goes off the rails, but it's pretty weird:

I am able to reproduce with the following small tests:
```go
func TestNonAtomic(t *testing.T) {
        var schema int32 = -1

        t.Logf("schema = %d", schema)
        div := 1 << -schema

        if div != 2 {
                t.Fatalf("div expected 2, got: %d", div)
        }
}

func TestAtomic(t *testing.T) {
        var schema int32 = -1
        atomicSchema := atomic.LoadInt32(&schema)

        t.Logf("atomicSchema = %d", atomicSchema)
        div := 1 << -atomicSchema

        if div != 2 {
                t.Fatalf("div expected 2, got: %d", div)
        }
}
```
```
=== RUN   TestNonAtomic
    main_test.go:11: schema = -1
--- PASS: TestNonAtomic (0.00s)
=== RUN   TestAtomic
    main_test.go:23: atomicSchema = -1
    main_test.go:27: div expected 2, got: 0
--- FAIL: TestAtomic (0.00s)
FAIL
FAIL    dswarbrick/foo  0.002s
FAIL
```
Both of the above tests obviously pass (as expected) on amd64.

And more from the "strange but true" department, `atomic.LoadInt64` behaves as expected:
```go
func TestAtomicInt32(t *testing.T) {
        var schema int32 = -1
        atomicInt32 := atomic.LoadInt32(&schema)

        t.Logf("atomicInt32 = %d", atomicInt32)
        div := 1 << -atomicInt32
        t.Logf("div = %d", div)

        if div != 2 {
                t.Fatalf("div expected 2, got: %d", div)
        }
}

func TestAtomicInt64(t *testing.T) {
        var schema int64 = -1
        atomicInt64 := atomic.LoadInt64(&schema)

        t.Logf("atomicInt64 = %d", atomicInt64)
        div := 1 << -atomicInt64
        t.Logf("div = %d", div)

        if div != 2 {
                t.Fatalf("div expected 2, got: %d", div)
        }
}
```
```
=== RUN   TestAtomicInt32
    main_test.go:12: atomicInt32 = -1
    main_test.go:14: div = 0
    main_test.go:17: div expected 2, got: 0
--- FAIL: TestAtomicInt32 (0.00s)
=== RUN   TestAtomicInt64
    main_test.go:25: atomicInt64 = -1
    main_test.go:27: div = 2
--- PASS: TestAtomicInt64 (0.00s)
FAIL
FAIL    dswarbrick/foo  0.002s
FAIL
```
`TestAtomicInt32` also passes if I cast to int64, e.g.
```go
div := 1 << int64(-atomicInt32)
```
This has surely got to be a compiler bug.

--- a/prometheus/histogram.go
+++ b/prometheus/histogram.go
@@ -544,16 +544,12 @@ func newHistogram(desc *Desc, opts HistogramOpts, labelValues ...string) Histogr
 	}
 	// Finally we know the final length of h.upperBounds and can make buckets
 	// for both counts as well as exemplars:
-	h.counts[0] = &histogramCounts{
-		buckets:                          make([]uint64, len(h.upperBounds)),
-		nativeHistogramZeroThresholdBits: math.Float64bits(h.nativeHistogramZeroThreshold),
-		nativeHistogramSchema:            h.nativeHistogramSchema,
-	}
-	h.counts[1] = &histogramCounts{
-		buckets:                          make([]uint64, len(h.upperBounds)),
-		nativeHistogramZeroThresholdBits: math.Float64bits(h.nativeHistogramZeroThreshold),
-		nativeHistogramSchema:            h.nativeHistogramSchema,
-	}
+	h.counts[0] = &histogramCounts{buckets: make([]uint64, len(h.upperBounds))}
+	atomic.StoreUint64(&h.counts[0].nativeHistogramZeroThresholdBits, math.Float64bits(h.nativeHistogramZeroThreshold))
+	atomic.StoreInt32(&h.counts[0].nativeHistogramSchema, h.nativeHistogramSchema)
+	h.counts[1] = &histogramCounts{buckets: make([]uint64, len(h.upperBounds))}
+	atomic.StoreUint64(&h.counts[1].nativeHistogramZeroThresholdBits, math.Float64bits(h.nativeHistogramZeroThreshold))
+	atomic.StoreInt32(&h.counts[1].nativeHistogramSchema, h.nativeHistogramSchema)
 	h.exemplars = make([]atomic.Value, len(h.upperBounds)+1)
 
 	h.init(h) // Init self-collection.

--- a/prometheus/histogram_test.go
+++ b/prometheus/histogram_test.go
@@ -470,7 +470,7 @@ func TestHistogramExemplar(t *testing.T) {
 	}
 }
 
-func TestSparseHistogram(t *testing.T) {
+func TestNativeHistogram(t *testing.T) {
 	scenarios := []struct {
 		name             string
 		observations     []float64 // With simulated interval of 1m.
@@ -686,7 +686,7 @@ func TestSparseHistogram(t *testing.T) {
 	}
 }
 
-func TestSparseHistogramConcurrency(t *testing.T) {
+func TestNativeHistogramConcurrency(t *testing.T) {
 	if testing.Short() {
 		t.Skip("Skipping test in short mode.")
 	}
@@ -703,7 +703,7 @@ func TestSparseHistogramConcurrency(t *testing.T) {
 		end.Add(concLevel)
 
 		his := NewHistogram(HistogramOpts{
-			Name:                            "test_sparse_histogram",
+			Name:                            "test_native_histogram",
 			Help:                            "This help is sparse.",
 			NativeHistogramBucketFactor:     1.05,
 			NativeHistogramZeroThreshold:    0.0000001,

